server/cluster/metadata/cluster_metadata.go

```diff
@@ -402,6 +406,32 @@ func (c *ClusterMetadata) RegisterNode(ctx context.Context, registeredNode Regis
 	return nil
 }
 
+func needUpdate(oldCache RegisteredNode, registeredNode RegisteredNode) bool {
+	oldShardIDs := make([]storage.ShardID, 0, len(oldCache.ShardInfos))
+	for i := 0; i < len(oldCache.ShardInfos); i++ {
+		oldShardIDs = append(oldShardIDs, oldCache.ShardInfos[i].ID)
+	}
+	sort.Slice(oldShardIDs, func(i, j int) bool {
+		return oldShardIDs[i] < oldShardIDs[j]
+	})
+	curShardIDs := make([]storage.ShardID, 0, len(registeredNode.ShardInfos))
+	for i := 0; i < len(registeredNode.ShardInfos); i++ {
+		curShardIDs = append(curShardIDs, registeredNode.ShardInfos[i].ID)
+	}
+	sort.Slice(curShardIDs, func(i, j int) bool {
+		return curShardIDs[i] < curShardIDs[j]
+	})
+	if len(oldShardIDs) != len(curShardIDs) {
+		return true
+	}
+	for i := 0; i < len(curShardIDs); i++ {
+		if curShardIDs[i] != oldShardIDs[i] {
+			return true
+		}
+	}
```

Comment:
Maybe it is better to do a two level forloop to check whether the shard infos are the same to avoid any sort and memory allocation.